src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,407 @@
+            }
+            return null;
+        }
+
+        /*
+        Increment total hit count and validate if threshold is reached.
+         */
+        protected void incrementTotalHitCount() throws IOException {
+            totalHits++;
+            hitsThresholdChecker.incrementHitCount();
+            if (scoreMode().isExhaustive() == false
+                && getTotalHitsRelation() == TotalHits.Relation.EQUAL_TO
+                && hitsThresholdChecker.isThresholdReached()) {
+                // for the first time hitsThreshold is reached, notify all comparators about this
+                for (LeafFieldComparator comparator : comparators) {
+                    comparator.setHitsThresholdReached();
+                }
+                setTotalHitsRelation(TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO);
+            }
+        }
+
+        /*
+        Collect hit and add the value of the sort field in the comparator.
+         */
+        protected void collectHit(int doc, int hitsCollected, int subQueryNumber, float score) throws IOException {
+            // Startup transient: queue hasn't gathered numHits yet
+            int slot = hitsCollected - 1;
+            // Copy hit into queue
+            if (numHits > 0) {
+                comparators[subQueryNumber].copy(slot, doc);
+                add(slot, doc, compoundScores[subQueryNumber], subQueryNumber, score);
+                if (queueFull[subQueryNumber]) {
+                    comparators[subQueryNumber].setBottom(bottom.slot);
+                }
+            } else {
+                queueFull[subQueryNumber] = true;
+            }
+        }
+
+        /*
+         * This hit is competitive - replace bottom element in queue & adjustTop
+         */
+        protected void collectCompetitiveHit(int doc, int subQueryNumber) throws IOException {
+            // This hit is competitive - replace bottom element in queue & adjustTop
+            if (numHits > 0) {
+                comparators[subQueryNumber].copy(bottom.slot, doc);
+                updateBottom(doc, compoundScores[subQueryNumber]);
+                comparators[subQueryNumber].setBottom(bottom.slot);
+            }
+        }
+
+        protected boolean thresholdCheck(int doc, int subQueryNumber) throws IOException {
+            if (collectedAllCompetitiveHits || reverseMul * comparators[subQueryNumber].compareBottom(doc) <= 0) {
+                // since docs are visited in doc Id order, if compare is 0, it means
+                // this document is larger than anything else in the queue, and
+                // therefore not competitive.
+                if (searchSortPartOfIndexSort) {
+                    if (hitsThresholdChecker.isThresholdReached()) {
+                        setTotalHitsRelation(TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO);
+                        log.info("Terminating collection as hits threshold is reached");
+                        throw new CollectionTerminatedException();
```

Comment:
why this exception? I don't see a code where we're handling this exception type in some special way. I would rather throw IllegalState where I can put some meaningful error message that gives some clues to users and engineers while troubleshooting.
Also why logging is at info level, if we're throwing runtime exception it should be an error level.

Reply:
In opensearch this exception is caught https://github.com/opensearch-project/OpenSearch/blob/0d01d1755e282763e5ea020112617ba25459863c/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L328

Reply:
Also this is not an error https://github.com/opensearch-project/OpenSearch/blob/0d01d1755e282763e5ea020112617ba25459863c/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L329 Check this comment out